blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md

```diff
@@ -0,0 +1,446 @@
+	mapping(uint => Listing) private _listings;
+
+	function listToken(address token, uint tokenId, uint price) external {
+		IERC721(token).transferFrom(msg.sender, address(this), tokenId);
+
+		Listing memory listing = Listing(
+			payable(msg.sender),
+            payable(address(this)),
+			token,
+			tokenId,
+			price,
+            false
+		);
+		_listings[_listingId] = listing;
+		_listingId++;
+
+	}
+
+	function getListing(uint listingId) public view returns (Listing memory) {
+		return _listings[listingId];
+	}
+
+    function getListingLength()public view returns (uint){
+        return _listingId;
+    }
+
+	function buyToken(uint listingId) public payable {
+		Listing storage listing = _listings[listingId];
+
+		require(msg.value == listing.price, "Insufficient payment");
+
+		IERC721(listing.token).transferFrom(listing.owner, msg.sender, listing.tokenId);
+		payable(listing.seller).transfer(msg.value);
+
+        listing.sold = true;
+        listing.price = listing.price * 2;
+        listing.owner = payable(msg.sender);
+		listing.seller = payable(msg.sender);
+
+	}
+}
+```
+
+## Writing Tests
+
+Before you begin writing tests, you would create a folder specific for tests, in our case, it would be called `tests`then create a javascript file and name it whatever you want but in this case it would be `dripto-tests.js`.
+
+Firstly, you would import the necessary functions from the required modules.
+
+```js
+const { expect } = require("chai");
+const { ethers } = require("hardhat");
+```
+
+Here's a breakdown of each line:
+
+- `const { expect } = require("chai");`: This line imports the `expect` assertion function from the Chai library using the `require` function. The `expect` function is used to define the expected behavior of a function or code block in a test case.
+
+- `const { ethers } = require("hardhat");`: This line imports the `ethers` library from Hardhat. `ethers` is a JavaScript library that provides a clean, simple, and efficient way to interact with the Ethereum or Celo blockchain.
+
+Together, these two lines of code import the necessary libraries for writing and testing smart contracts in a Hardhat project. They allow developers to use Chai's assertion functions and Hardhat's `ethers` library to write effective and efficient test cases for their smart contracts.
+
+It's worth noting that the use of `const` indicates that these variables are declared as constants and cannot be reassigned once they are initialized. This helps ensure the stability and consistency of the testing environment.
+
+Next, you are going to define two test suites and set a longer timeout for the test cases in the `MyNFT` suite. This allows the test cases in the `MyNFT` suite to take longer to execute, which may be necessary for more complex test cases or for cases where the blockchain network is slow to respond. 
+
+```js
+describe("MyNftMarket", function () {});
+
+describe("MyNFT", function () {
+  this.timeout(50000);
+```
```

Comment:
```suggestion
```js
describe("MyNftMarket", function () {});

describe("MyNFT", function(){
    this.timeout(50000)
    // The next code is below
})
```
```